scipy/optimize/_linprog_util.py

```diff
@@ -368,84 +374,74 @@ def _clean_inputs(lp):
-                and np.isreal(bounds[1])):
-            bounds = [(bounds[0], bounds[1])] * n_x
-        else:
-            raise ValueError(
-                "Invalid input for linprog: exactly one lower bound and one "
-                "upper bound must be specified for each element of x")
-
-        clean_bounds = []  # also creates a copy so user's object isn't changed
-        for i, b in enumerate(bounds):
-            if b[0] is not None and b[1] is not None and b[0] > b[1]:
-                raise ValueError(
-                    "Invalid input for linprog, bound " +
-                    str(i) +
-                    " " +
-                    str(b) +
-                    ": a lower bound must be less than or equal to the "
-                    "corresponding upper bound")
-            if b[0] == np.inf:
-                raise ValueError(
-                    "Invalid input for linprog, bound " +
-                    str(i) +
-                    " " +
-                    str(b) +
-                    ": infinity is not a valid lower bound")
-            if b[1] == -np.inf:
-                raise ValueError(
-                    "Invalid input for linprog, bound " +
-                    str(i) +
-                    " " +
-                    str(b) +
-                    ": negative infinity is not a valid upper bound")
-            lb = float(b[0]) if b[0] is not None and b[0] != -np.inf else None
-            ub = float(b[1]) if b[1] is not None and b[1] != np.inf else None
-            clean_bounds.append((lb, ub))
-        bounds = clean_bounds
-    except ValueError as e:
-        if "could not convert string to float" in e.args[0]:
-            raise TypeError
-        else:
-            raise e
-    except TypeError as e:
-        print(e)
-        raise TypeError(
-            "Invalid input for linprog: bounds must be a sequence of "
-            "(min,max) pairs, each defining bounds on an element of x ")
-
-    return _LPProblem(c, A_ub, b_ub, A_eq, b_eq, bounds, x0)
+    # Bounds can be one of these formats:
+    # (1) None
+    # (2) a sequence with 2 scalars
+    # (3) a sequence with 1 element as (2)
+    # (4) a sequence with N elements, all with 2 values (N is the size of x)
+    # (5) a sequence with 2 elements, both sequences with N values
+    # (6) a 2-D array, with shape N x 2 or 2 x N
+    # Unspecified bounds can be represented by None or (-)np.inf.
+    # All formats are converted into a N x 2 np.array with (-)np.inf where bounds are unspecified.
+    # Strings in input result in a ValueError
+    clean_bounds = np.zeros((n_x,2))
+
+    bounds_valid = False
+    # Determine shape of provided bounds
+    # np.shape returns a tuple, but only if sizes are consistent
+    bsh = np.shape(bounds)
+    
+    # 1. Check if bounds can be interpreted as n_x pairs (n_x is the number of variables)
+    #    Bounds can have sizes n_x x 2 and 2 x n_x
+    if len(bsh) == 2:
```

Comment:
I think checking dimensions should be done first. I thought the np.shape() functions would be OK for that (as in the PR), but perhaps not, because it does not catch all inconsistent dimensions (e.g. for `[[1, 2], [1, 2, 3]]` it returns `(2, )` which at first seemed manageable but after some more thought I'd prefer to have a exception thrown here.
I took a look at this today. A better solution may be `b = np.array(bounds,dtype=float); bsh = b.shape`, which not just raises an error if dimensions of bounds are inconsistent, but also if the input cannot be converted (both ValueErrors...). The array `b` can be used for further processing if it is not Nx2; transposing, for example.

Reply:
Sounds good. (Although I think transposing will never be necessary now that we're not accepting 2 x N.)

Reply:
See yesterday's update. Quite a bit tighter.